File: codeSignal.js
/* add
Write a function that returns the sum of two numbers. */
function solution(param1, param2) {
    return param1 + param2;
};

/* centuryFromYear
Given a year, return the century it is in. 
The first century spans from the year 1 up to and including the year 100, 
the second - from the year 101 up to and including the year 200, etc. */
function solution(year) {
    return Math.ceil(year/100);
};

/* checkPalindrome
Given the string, check if it is a palindrome. */
function solution(inputString) {
    for(let i = 0; i < inputString.length / 2; i++) {
        if(inputString[i] != inputString[inputString.length - 1 - i]) {
            return false;
        }
    }
    return true;
};

/* adjascentElementsProduct
Given an array of integers, 
find the pair of adjacent elements that has the largest product 
and return that product. */
function adjascentElementsProduct(inputArray) {
    let num = 0;
    for(let i = 0; i < inputArray.length-1; i++) {
        let x = inputArray[i+1];
        let product = inputArray[i] * x;
        if(product > num || i == 0) {
            num = product;
        }
    }
    return num;
};

/* shapeArea
Find the area of a polygon for a given n.
Ex: a 1-interesting polygon is just a square with a side of length 1.
An n-interesting polygon is obtained by taking the n - 1-interesting polygon
and appending 1-interesting polygons to its rim, side by side. (See CodeSignal
for example images.) */
function shapeArea(n) {
    // multiples of 4
    // loop going up n times
    let result = 1;
    if (n === 1) { // === also checks data type, better for debugging
        return result;
    }
    for (let i = 0; i < n; i++) {
        result += i*4;
        console.log("RESULT:",result);
    }
    return result;
};

/* Make Array Consecutive 2
Arrange statues from smallest to largest so that each statue will be bigger than the previous one by exactly 1.
May need additional statues to accomplish this. Determine min num of additional statues needed. */
function makeArrConsecutive(statues) { // statues = arr of +ints
    //sort statues arr in ascending order
    statues.sort(function(a,b){return a - b});
    // placeholder variable to be returned
    let count = 0;
    //iterate through sorted arr
    for(let i = 0; i < statues.length-1; i++) {
        if(statues[i] - statues[i+1] != -1) {
            count += statues[i+1] - statues[i] - 1;
        }
    }
    return count;
}
console.log(makeArrConsecutive([6,2,3,8]))

File: codeSignal/codeSignal.py
# Add
# Write a function that returns the sum of two numbers
def solution(param1, param2):
    return param1 + param2

# centuryFromYear
# Given a year, return the century it is in. The 1st century spans from year 1 up to & including year 100, 2nd century from year 101
# up to & including year 200, etc.
import math
def solution(year):
    century = math.ceil(year/100)
    return century

# checkPalindrome
# Given the string, check if it is a palindrome.
def solution(inputString):
    reversed_string = inputString[::-1] # slice statement: start at end of string, end at pos 0, move w/ step -1
    return reversed_string == inputString

# adjacentElementsProduct
# Given an arr of ints, find pair of adjacent elements w/ largest product & return that product.
def solution(inputArray):
    product_list = []
    for i in range(len(inputArray)-1):
        product = inputArray[i]*inputArray[i+1]
        product_list.append(product)
    product_list.sort()
    return product_list[-1]

# shapeArea
# Find the area of a polygon for a given n.
# Ex: a 1-interesting polygon is just a square with a side of length 1.
# An n-interesting polygon is obtained by taking the n - 1-interesting polygon
# and appending 1-interesting polygons to its rim, side by side. (See CodeSignal
# for example images.)
def solution(n):
    result = 1
    if n == 1:
        return result
    i = 0
    while i < n:
        result += (i*4)
        i+=1
    return result

# Make Array Consecutive 2
# Arrange given list from smallest to largest. Insert missing numbers to get a list of numbers which each increment by 1.
# Return how many numbers needed to be added to the list to accomplish this.
def solution(statues):
    statues.sort()
    count = 0
    i = 0
    while i < len(statues)-1:
        if statues[i+1] != statues[i]+1:
            statues.insert(i+1,statues[i]+1)
            count+=1
        i+=1
    return count

# almostIncreasingSequence
# Given a sequence of ints as an arr, determine whether it is possible to obtain a strictly increasing sequence by removing
# no more than 1 element from the arr. Function returns true or false.
# def solution9(sequence):
#     count = 0
#     i = 1
#     while i <= len(sequence):
#         if len(sequence) == i:
#             i-=1
#             if sequence[i-1] >= sequence[i]-1:
#                 sequence.pop(i)
#                 count+=1
#                 if count > 1:
#                     return False
#             break
#         if sequence[i-1] == sequence[i] or sequence[i-1] > sequence[i]-1 or sequence[i-2] > sequence[i-1]:
#             sequence.pop(i)
#             count+=1
#             if count > 1:
#                 return False
#         i+=1
#     return True
# print(solution9([1,2,5,3,5])) # True

def solution10(sequence):
    count = 0
    i = 0
    while i <= len(sequence)-1:
        if len(sequence)-1 == i:
            i-=1
            if sequence[i+1] <= sequence[i]:
                sequence.pop(i+1)
                count+=1
                if count > 1:
                    return False
        if i > 0 and sequence[i-1] >= sequence[i] or sequence[i+1] <= sequence[i]: # does not consider if sequence[0]>sequence[1]
            sequence.pop(i+1)
            count+=1
        if count > 1:
            return False
        i+=1
    return True
print(solution10([1,2,1,2]))